sycl/include/sycl/detail/impl_utils.hpp

```diff
@@ -0,0 +1,55 @@
+//===----------------------------------------------------------------------===//
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <sycl/detail/defines_elementary.hpp>
+
+#include <cassert>
+
+namespace sycl {
+__SYCL_INLINE_VER_NAMESPACE(_V1) {
+namespace detail {
+
+// Helper function for extracting implementation from SYCL's interface objects.
+// Note! This function relies on the fact that all SYCL interface classes
+// contain "impl" field that points to implementation object. "impl" field
+// should be accessible from this function.
+//
+// Note that due to a bug in MSVC compilers (including MSVC2019 v19.20), it
+// may not recognize the usage of this function in friend member declarations
+// if the template parameter name there is not equal to the name used here,
+// i.e. 'Obj'. For example, using 'Obj' here and 'T' in such declaration
+// would trigger that error in MSVC:
+//   template <class T>
+//   friend decltype(T::impl) detail::getSyclObjImpl(const T &SyclObject);
+template <class Obj> decltype(Obj::impl) getSyclObjImpl(const Obj &SyclObject) {
+  assert(SyclObject.impl && "every constructor should create an impl");
+  return SyclObject.impl;
+}
+
+// Returns the raw pointer to the impl object of given face object. The caller
+// must make sure the returned pointer is not captured in a field or otherwise
+// stored - i.e. must live only as on-stack value.
+template <class T>
+typename std::add_pointer_t<typename decltype(T::impl)::element_type>
+getRawSyclObjImpl(const T &SyclObject) {
+  return SyclObject.impl.get();
+}
+
+// Helper function for creation SYCL interface objects from implementations.
+// Note! This function relies on the fact that all SYCL interface classes
+// contain "impl" field that points to implementation object. "impl" field
+// should be accessible from this function.
+template <class T> T createSyclObjFromImpl(decltype(T::impl) ImplObj) {
+  return T(ImplObj);
+}
```

Comment:
Let's assume `T == event` for the context of this comment. From what I understand, the only purpose of this interface is to avoid

```
class event {
public:
  event(sycl::detail::event_impl) {...}
};
```

by making that ctor protected/private and using this helper whenever we want to create `event` objects from the `event_impl`. Do we really need to do that to comply with the standard?

Reply:
CC-ing @romanovvlad, who might know the origins of this design decision.

I don't think the standard explicitly restricts extending SYCL API with undocumented member functions, nor that the standard requirements are the only design drivers for DPC++ API. I suggest starting a separate issue to discuss this refactoring. I hesitate to make such changes within this PR.

Reply:
The reason I'm asking it here is because if two out of three are/can be made redundant, then the first one could be changed from a template to overloads where each overload is defined as part of inline friend declaration/definition, eliminating the need for this header alltogether.

Reply:
This header quite is small comparing to STL header included in common.hpp. I haven't evaluated your suggestions yet, but if they add runtime overhead it might be not worth it.

Reply:
>> by making that ctor protected/private and using this helper whenever we want to create event objects from the event_impl. 

Right, the helper is for avoiding non-standard public method which would reference internal types.